### PR TITLE
Hide enum types in the stdlib and the manual

### PR DESCRIPTION
**Pull request by yannham**
The enum type syntax has been changed temporarily since the unified syntax
change. While we bikeshed on a new solid syntax, we don't want users to rely
on the temporary one.

This PR:

- Hides enum types from stdlib types by wrapping them as contracts with meaningful names and error messages.  When a user write, say, ``builtin.serialize `NonExistentFormat "foo"``, the error they will face is understandable without using the enum type syntax.
- Documents enum tags but not enum types in the syntax section of the manual.
- Hide documentation about enum types in the typing section of the manual.

That way, they can still be used under the hood the stdlib, but without facing users.

### CODE REVIEW

File: doc/manual/syntax.md [hunk 1]
```diff
@@ -178,6 +179,40 @@ def concat(str_array, log=false):
...
+#       "
+
+```
+
+While it's technically possible to just use strings in place of enum tags, using
+an enum tag insists on the fact that only a finite number of alternatives can be
+used for the corresponding value.
+
+Enum are also handled more finely by the typechecker, however, as of the version
+0.1 of Nickel, enum types have been temporary disabled for incidental reasons.
```
Comment (aspiwack) [inline doc/manual/syntax.md:214]: Be more specific: “there are only two enum types[…]”

File: stdlib/array.ncl [hunk 2]
```diff
@@ -254,7 +269,7 @@
         "%m
       = fun f n => %generate% n f,
 
-    sort : forall a. (a -> a -> [| Less, Equal, Greater |]) -> Array a -> Array a
+    sort : forall a. (a -> a -> ComparisonResult) -> Array a -> Array a
```
Comment (aspiwack) [inline stdlib/array.ncl:272]: Is it not statically typechecked? The (static) type checker currently reads this as an opaque type, why doesn't it complain?

Comment (yannham) [reply stdlib/array.ncl:272]: This is why the CI is failing currently, I missed this one. Indeed we need a contract application.

File: doc/manual/syntax.md (hunk 1 above)
Comment (yannham) [reply doc/manual/syntax.md:214]: You mean, the one from the stdlib? Those are standard custom contracts. I wrote them using enum types just because I can, because it's more concise and doesn't cost anything, but any user can write:

```nickel
MyComparisonResult = contract.from_predicate (fun value => value == `Less || value == `Equal || value == `Greater)
```

Should we really mention them as "enum types" in the doc, although they are just opaque types, statically?

File: stdlib/array.ncl (hunk 2 above)
Comment (aspiwack) [reply stdlib/array.ncl:272]: I don't understand your answer, I'm afraid :slightly_smiling_face:

File: doc/manual/syntax.md (hunk 1 above)
Comment (aspiwack) [reply doc/manual/syntax.md:214]: I changed my mind, now that we've discussed the details. Let's say

```suggestion
Enum will also handled more finely by the typechecker, however, as of the version
0.1 of Nickel, enum types are not yet supported.
```

**Review by aspiwack — APPROVED** [review]
I see. Good application of gradual typing.

Honestly, I would have rather we hid the syntax of enum types completely, and implemented `ComparisonResult` & co as magic constants. But it's probably the more pragmatic choice.

Maybe label every use with a comment saying “internal syntax, don't rely on it”.
